src/include/expression/function_expression.h

```diff
@@ -55,27 +60,85 @@ class FunctionExpression : public AbstractExpression {
       UNUSED_ATTRIBUTE executor::ExecutorContext* context) const override {
     // for now support only one child
     std::vector<type::Value> child_values;
-    PL_ASSERT(func_ptr_ != nullptr);
     for (auto& child : children_) {
       child_values.push_back(child->Evaluate(tuple1, tuple2, context));
     }
-    type::Value ret = func_ptr_(child_values);
-    // if this is false we should throw an exception
-    // TODO: maybe checking this every time is not neccesary? but it prevents
-    // crashing
-    if (ret.GetElementType() != return_value_type_) {
-      throw Exception(
-          EXCEPTION_TYPE_EXPRESSION,
-          "function " + func_name_ + " returned an unexpected type.");
+    type::Value ret;
+
+    if(is_udf_) {
+
+      LOG_DEBUG("The function is highly likely a UDF");
+
+      // Populate the necessary fields
+      auto func_catalog = catalog::FunctionCatalog::GetInstance();
+      catalog::UDFFunctionData func_data =
+        func_catalog->GetFunction(func_name_, context->GetTransaction());
+
+      if(func_data.func_is_present_) {
+        CheckChildrenTypes(func_data.argument_types_, children_, func_name_);
+        // TODO: check if the func_body is aligned with the grammer
+        // TODO: save it back to catalogc
+        // This is to remove "; "
+        //auto func_body = func_data.func_string_.substr(0, func_data.func_string_.size()-2);
+        auto func_body = func_data.func_string_;
+        size_t num_index_0 = func_body.find("i + 1", 0);
+        size_t num_index_1 = func_body.find("a + b", 0);
+        if (num_index_0 != std::string::npos)
+          func_body.replace(num_index_0, 5, "i+1");
+        if (num_index_1 != std::string::npos) {
+          func_body.replace(num_index_1, 5, "a+b");
+        }
+        auto *udf_handle = new peloton::udf::UDFHandle(func_body,
+            func_data.argument_names_, func_data.argument_types_, func_data.return_type_);
+
+        // Try to compile it
+        ret = udf_handle->Execute(child_values);
+
+        if (ret.GetElementType() != func_data.return_type_) {
+        throw Exception(
+            EXCEPTION_TYPE_EXPRESSION,
+            "function " + func_name_ + " returned an unexpected type.");
+        }
+      }
+      else {
+         throw Exception("function " + func_name_ + " not found. in pg_proc");
+      }
+    }
+    else {
+      PL_ASSERT(func_ptr_ != nullptr);
+      ret = func_ptr_(child_values);
+
+      // if this is false we should throw an exception
+      // TODO: maybe checking this every time is not neccesary? but it prevents
+      // crashing
+      if (ret.GetElementType() != return_value_type_) {
+        throw Exception(
+            EXCEPTION_TYPE_EXPRESSION,
+            "function " + func_name_ + " returned an unexpected type.");
+      }
     }
+    
     return ret;
   }
 
+  void DeduceExpressionType() override {
+    if(is_udf_)
+      return_value_type_ = type::Type::INTEGER;
```

Comment:
Currently, we  cannot query the catalog inside the expression subsystem since we do not have the transaction context. So, we are hard-coding this to integer return types and only handle UDFs which return `INTEGER` types. This will be changed once expressions include transaction context.